lib/gemstash/gem_source.rb

```diff
@@ -0,0 +1,89 @@
+require "gemstash"
+require "forwardable"
+
+module Gemstash
+  #:nodoc:
+  module GemSource
+    autoload :PrivateSource,  "gemstash/gem_source/private_source"
+    autoload :RackMiddleware, "gemstash/gem_source/rack_middleware"
+    autoload :RedirectSource, "gemstash/gem_source/upstream_source"
+    autoload :RubygemsSource, "gemstash/gem_source/upstream_source"
+    autoload :UpstreamSource, "gemstash/gem_source/upstream_source"
+
+    API_REQUEST_LIMIT = 200
+
+    def self.sources
+      @sources ||= [
+        Gemstash::GemSource::PrivateSource,
+        Gemstash::GemSource::RedirectSource,
+        Gemstash::GemSource::UpstreamSource,
+        Gemstash::GemSource::RubygemsSource
```

Comment:
I think that the only difference between an UpstreamSource and the RubygemsSource is the upstream endpoint. Does it make sense to handle that complexity inside an *upstream* object?